File: apps/ios/Pebbles/Features/Path/Render/PebbleSVGModel.swift
import CoreGraphics
import Foundation
import os

/// Typed view of a composed pebble SVG: viewBox + ordered layers, each carrying
/// its `transform` chain, opacity, and a single combined `CGPath` of all the
/// descendant `<path d="…">` strings concatenated together.
///
/// Built once on first appearance of `PebbleAnimatedRenderView`. Failure
/// (`init?` returning nil) means the caller falls back to the existing
/// `SVGView`-based static renderer.
struct PebbleSVGModel {
    let viewBox: CGRect
    let layers: [Layer]

    struct Layer {
        enum Kind { case shape, fossil, glyph }
        let kind: Kind
        /// Transform inherited from the layer's `<g transform="...">`. Identity if none.
        let transform: CGAffineTransform
        /// Layer opacity from `<g opacity="...">`. 1.0 if absent.
        let opacity: Double
        /// All descendant `<path>`s parsed and concatenated into one path.
        /// Any non-layer nested-group transforms are already baked in.
        let combinedPath: CGPath
    }

    init?(svg: String) {
        guard let data = svg.data(using: .utf8) else { return nil }
        let parser = XMLParser(data: data)
        let delegate = ParserDelegate()
        parser.delegate = delegate
        guard parser.parse(),
              !delegate.parseFailed,
              let viewBox = delegate.viewBox else {
            Logger(subsystem: "app.pbbls.ios", category: "pebble-svg")
                .info("PebbleSVGModel parse failed — falling back to static render")
            return nil
        }

        var layers: [Layer] = []
        for raw in delegate.rawLayers {
            guard !raw.paths.isEmpty else { continue }
            let combined = CGMutablePath()
            for path in raw.paths {
                combined.addPath(path)
            }
            guard !combined.boundingBoxOfPath.isNull else { continue }
            layers.append(Layer(
                kind: raw.kind,
                transform: raw.transform,
                opacity: raw.opacity,
                combinedPath: combined.copy() ?? combined
            ))
        }

        guard !layers.isEmpty else { return nil }
        self.viewBox = viewBox
        self.layers = layers
    }

    // MARK: - XMLParser delegate

    fileprivate struct RawLayer {
        let kind: Layer.Kind
        let transform: CGAffineTransform
        let opacity: Double
        let paths: [CGPath]
    }

    fileprivate struct RawGroup {
        var kind: Layer.Kind?
        var transform: CGAffineTransform = .identity
        var opacity: Double = 1.0
        // Paths attached to this <g>. Already parsed to CGPath. Inner-group
        // transforms are baked in as those groups close.
        var paths: [CGPath] = []
    }

    private final class ParserDelegate: NSObject, XMLParserDelegate {
        var viewBox: CGRect?
        var rawLayers: [RawLayer] = []
        var parseFailed = false
        private var stack: [RawGroup] = []

        func parser(
            _ parser: XMLParser,
            didStartElement elementName: String,
            namespaceURI: String?,
            qualifiedName qName: String?,
            attributes attributeDict: [String: String] = [:]
        ) {
            switch elementName {
            case "svg":
                if let viewBoxAttr = attributeDict["viewBox"] {
                    let parts = viewBoxAttr
                        .split(whereSeparator: { $0 == " " || $0 == "," })
                        .compactMap { Double($0) }
                    if parts.count == 4 {
                        viewBox = CGRect(x: parts[0], y: parts[1], width: parts[2], height: parts[3])
                    }
                }
            case "g":
                var group = RawGroup()
                if let id = attributeDict["id"] {
                    switch id {
                    case "layer:shape":  group.kind = .shape
                    case "layer:fossil": group.kind = .fossil
                    case "layer:glyph":  group.kind = .glyph
                    default: break
                    }
                }
                if let transformAttr = attributeDict["transform"] {
                    group.transform = parseTransform(transformAttr)
                }
                if let opacityAttr = attributeDict["opacity"], let value = Double(opacityAttr) {
                    group.opacity = value
                }
                stack.append(group)
            case "path":
                guard let dString = attributeDict["d"], !stack.isEmpty else { return }
                if let parsedPath = SVGPathParser.parse(dString) {
                    stack[stack.count - 1].paths.append(parsedPath)
                } else {
                    parseFailed = true
                }
            default:
                break
            }
        }

        func parser(
            _ parser: XMLParser,
            didEndElement elementName: String,
            namespaceURI: String?,
            qualifiedName qName: String?
        ) {
            guard elementName == "g", let group = stack.popLast() else { return }

            if let kind = group.kind {
                // Register the layer. Its paths already have any inner-group
                // transforms baked in; the layer's own `transform` is applied
                // by the renderer (LayerShape).
                rawLayers.append(RawLayer(
                    kind: kind,
                    transform: group.transform,
                    opacity: group.opacity,
                    paths: group.paths
                ))
            } else if !stack.isEmpty {
                // Non-layer nested group: bake this group's transform into
                // each of its paths and propagate them upward to the parent.
                var transform = group.transform
                for path in group.paths {
                    if let transformed = path.copy(using: &transform) {
                        stack[stack.count - 1].paths.append(transformed)
                    } else {
                        stack[stack.count - 1].paths.append(path)
                    }
                }
            }
            // Else: orphan non-layer group at the root with no parent — discard.
        }

        /// Parses the limited transform forms emitted by the engine:
        /// `translate(x, y) scale(s)` (commas optional), `translate(x, y)`,
        /// `scale(s)`. Other forms collapse to identity.
        ///
        /// Composition matches SVG semantics: the transform list applies
        /// left-to-right to a point, i.e. for `translate(40,40) scale(0.8)`
        /// the point is first scaled and then translated. In CG row-vector
        /// math (`p' = p * M`), that means each subsequent op is post-
        /// multiplied onto the running matrix via `op.concatenating(result)`.
        private func parseTransform(_ string: String) -> CGAffineTransform {
            var result = CGAffineTransform.identity
            let pattern = #"(translate|scale)\s*\(([^)]*)\)"#
            guard let regex = try? NSRegularExpression(pattern: pattern) else { return .identity }
            let range = NSRange(string.startIndex..<string.endIndex, in: string)
            regex.enumerateMatches(in: string, range: range) { match, _, _ in
                guard
                    let match,
                    let nameRange = Range(match.range(at: 1), in: string),
                    let argsRange = Range(match.range(at: 2), in: string)
                else { return }
                let name = String(string[nameRange])
                let args = String(string[argsRange])
                    .split(whereSeparator: { $0 == "," || $0 == " " })
                    .compactMap { Double($0) }
                switch name {
                case "translate":
                    let tx = args.first ?? 0
                    let ty = args.count > 1 ? args[1] : 0
                    result = CGAffineTransform(translationX: tx, y: ty).concatenating(result)
                case "scale":
                    let sx = args.first ?? 1
                    let sy = args.count > 1 ? args[1] : sx
                    result = CGAffineTransform(scaleX: sx, y: sy).concatenating(result)
                default:
                    break
                }
            }
            return result
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/render/PebbleSvgModel.kt
package app.pbbls.android.features.path.render

import org.xml.sax.Attributes
import org.xml.sax.InputSource
import org.xml.sax.helpers.DefaultHandler
import java.io.StringReader
import javax.xml.parsers.SAXParserFactory

/**
 * Typed, render-agnostic view of a server-composed pebble SVG: the viewBox plus
 * the ordered `layer:*` groups, each carrying its opacity and its descendant
 * `<path>`s (path data + the flattened `translate`/`scale` transform inherited
 * from every enclosing `<g>`). The Android analog of iOS `PebbleSVGModel`.
 *
 * Deliberately pure — no Compose or `android.graphics` types — so the parse can
 * be unit-tested on the plain JVM (no Robolectric). [PebbleStaticRender] turns
 * this into Compose paths and traces every layer at [PebbleStroke.OUTLINE_WIDTH].
 *
 * [parsePebbleSvg] returns `null` when the markup has no viewBox or no traceable
 * layer, so the caller can fall back to the raw AndroidSVG renderer ([PebbleSvg]).
 */
internal data class PebbleSvgModel(
    val viewBox: ViewBox,
    val layers: List<Layer>,
) {
    /** Parsed `viewBox="minX minY width height"`. */
    data class ViewBox(
        val minX: Float,
        val minY: Float,
        val width: Float,
        val height: Float,
    )

    /** One `layer:*` group. [opacity] is the `<g opacity>` (1 when absent). */
    data class Layer(
        val opacity: Float,
        val paths: List<PathSpec>,
    )

    /**
     * One traceable `<path>`: its `d` data and the [transform] flattened from
     * every enclosing `<g transform>` (identity when none). Stroke width is
     * deliberately dropped — the renderer overrides it with the outline weight.
     */
    data class PathSpec(
        val d: String,
        val transform: Affine,
    )
}

/**
 * A 2-D affine transform limited to the `translate`/`scale` forms the engine
 * emits (matrix `[[a c e] [b d f] [0 0 1]]`). Maps a point `(x, y)` to
 * `(a·x + c·y + e, b·x + d·y + f)`.
 */
internal data class Affine(
    val a: Float,
    val b: Float,
    val c: Float,
    val d: Float,
    val e: Float,
    val f: Float,
) {
    /** `this ∘ other` — the transform that applies [other] first, then this. */
    fun concat(other: Affine): Affine =
        Affine(
            a = a * other.a + c * other.b,
            b = b * other.a + d * other.b,
            c = a * other.c + c * other.d,
            d = b * other.c + d * other.d,
            e = a * other.e + c * other.f + e,
            f = b * other.e + d * other.f + f,
        )

    companion object {
        val IDENTITY = Affine(1f, 0f, 0f, 1f, 0f, 0f)

        fun translate(
            tx: Float,
            ty: Float,
        ) = Affine(1f, 0f, 0f, 1f, tx, ty)

        fun scale(
            sx: Float,
            sy: Float,
        ) = Affine(sx, 0f, 0f, sy, 0f, 0f)
    }
}

/**
 * Parses [svg] into a [PebbleSvgModel], or `null` when it can't be traced
 * (malformed XML, missing viewBox, or no layer carries a stroked path).
 *
 * Uses SAX (available on both the JVM and Android) with a running stack of
 * enclosing-group transforms, mirroring the streaming walk in iOS
 * `PebbleSVGModel`. Paths that end up `fill="none"` **and** `stroke="none"` are
 * dropped: the engine emits such paths for shape-detail fills that `stripFills`
 * zeroed, and AndroidSVG skips them — tracing them would draw spurious strokes.
 */
internal fun parsePebbleSvg(svg: String): PebbleSvgModel? {
    val handler = PebbleSvgHandler()
    return try {
        val factory = SAXParserFactory.newInstance().apply { isNamespaceAware = false }
        // Defensive: the input is our own engine output, but keep the parser
        // from ever reaching for a DTD / external entity.
        runCatching { factory.setFeature(LOAD_EXTERNAL_DTD_FEATURE, false) }
        factory.newSAXParser().parse(InputSource(StringReader(svg)), handler)
        val viewBox = handler.viewBox ?: return null
        val layers = handler.layers.mapNotNull { it.toLayerOrNull() }
        if (layers.isEmpty()) null else PebbleSvgModel(viewBox, layers)
    } catch (e: Exception) {
        // Malformed markup is a runtime data condition, not a crash — the caller
        // falls back to AndroidSVG. (Broad catch: SAX can throw beyond
        // SAXException, e.g. IOException from the reader.)
        null
    }
}

// ── SAX handler ─────────────────────────────────────────────

private class MutableLayer(
    val opacity: Float,
) {
    val paths = mutableListOf<PebbleSvgModel.PathSpec>()

    fun toLayerOrNull(): PebbleSvgModel.Layer? = if (paths.isEmpty()) null else PebbleSvgModel.Layer(opacity, paths.toList())
}

/** One `<g>` frame: the flattened transform in effect, and the layer it feeds. */
private class GroupFrame(
    val transform: Affine,
    val layer: MutableLayer?,
)

private class PebbleSvgHandler : DefaultHandler() {
    var viewBox: PebbleSvgModel.ViewBox? = null
    val layers = mutableListOf<MutableLayer>()
    private val stack = ArrayDeque<GroupFrame>()

    override fun startElement(
        uri: String?,
        localName: String?,
        qName: String?,
        attributes: Attributes,
    ) {
        when (qName) {
            "svg" -> viewBox = parseViewBox(attributes.getValue("viewBox"))
            "g" -> {
                val parent = stack.lastOrNull()
                val parentTransform = parent?.transform ?: Affine.IDENTITY
                val own = attributes.getValue("transform")?.let(::parseTransform) ?: Affine.IDENTITY
                val transform = parentTransform.concat(own)
                val id = attributes.getValue("id")
                val layer =
                    if (id != null && id.startsWith("layer:")) {
                        val opacity = attributes.getValue("opacity")?.toFloatOrNull() ?: 1f
                        MutableLayer(opacity).also { layers.add(it) }
                    } else {
                        // A non-layer group (e.g. the glyph's inner normalization
                        // <g>) feeds whatever layer its parent belongs to.
                        parent?.layer
                    }
                stack.addLast(GroupFrame(transform, layer))
            }
            "path" -> {
                val frame = stack.lastOrNull() ?: return
                val layer = frame.layer ?: return
                val d = attributes.getValue("d") ?: return
                // SVG spec defaults: fill black, stroke none. A path left with
                // neither (a stripped shape-detail fill) is invisible — skip it.
                val fill = attributes.getValue("fill") ?: "black"
                val stroke = attributes.getValue("stroke") ?: "none"
                if (fill == "none" && stroke == "none") return
                layer.paths.add(PebbleSvgModel.PathSpec(d, frame.transform))
            }
        }
    }

    override fun endElement(
        uri: String?,
        localName: String?,
        qName: String?,
    ) {
        if (qName == "g") stack.removeLastOrNull()
    }
}

// ── Attribute parsing ───────────────────────────────────────

private const val LOAD_EXTERNAL_DTD_FEATURE = "http://apache.org/xml/features/nonvalidating/load-external-dtd"
private val ARG_SEPARATOR = Regex("[,\\s]+")
private val TRANSFORM_OP = Regex("(translate|scale)\\s*\\(([^)]*)\\)")

private fun parseViewBox(raw: String?): PebbleSvgModel.ViewBox? {
    if (raw == null) return null
    val parts = raw.trim().split(ARG_SEPARATOR).mapNotNull { it.toFloatOrNull() }
    if (parts.size != 4) return null
    return PebbleSvgModel.ViewBox(parts[0], parts[1], parts[2], parts[3])
}

/**
 * Parses the limited transform forms the engine emits —
 * `translate(x, y) scale(s)`, `translate(x, y)`, `scale(sx, sy)` (commas
 * optional) — composing left-to-right so the leftmost op is outermost (applied
 * last to a point), matching SVG semantics. Unrecognized forms collapse to
 * identity.
 */
internal fun parseTransform(value: String): Affine {
    var result = Affine.IDENTITY
    for (match in TRANSFORM_OP.findAll(value)) {
        val name = match.groupValues[1]
        val args = match.groupValues[2].split(ARG_SEPARATOR).mapNotNull { it.trim().toFloatOrNull() }
        val op =
            when (name) {
                "translate" -> Affine.translate(args.getOrElse(0) { 0f }, args.getOrElse(1) { 0f })
                "scale" -> {
                    val sx = args.getOrElse(0) { 1f }
                    Affine.scale(sx, args.getOrElse(1) { sx })
                }
                else -> Affine.IDENTITY
            }
        result = result.concat(op)
    }
    return result
}
